File: src/main/java/mc/rpgstats/advancemnents/LevelUpCriterion.java
package mc.rpgstats.advancemnents;

import com.google.gson.JsonObject;
import mc.rpgstats.main.CustomComponents;
import mc.rpgstats.main.RPGStats;
import net.minecraft.advancement.criterion.AbstractCriterion;
import net.minecraft.advancement.criterion.AbstractCriterionConditions;
import net.minecraft.predicate.entity.AdvancementEntityPredicateDeserializer;
import net.minecraft.predicate.entity.AdvancementEntityPredicateSerializer;
import net.minecraft.predicate.entity.EntityPredicate;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.util.Identifier;

public class LevelUpCriterion extends AbstractCriterion<LevelUpCriterion.LevelCriteria> {
    public static final Identifier ID = new Identifier(RPGStats.MOD_ID, "player_level");
    private static final Identifier ANY_ID = new Identifier(RPGStats.MOD_ID, "_any");
    
    @Override
    protected LevelCriteria conditionsFromJson(JsonObject obj, EntityPredicate.Extended pred, AdvancementEntityPredicateDeserializer predicateDeserializer) {
        return new LevelCriteria(pred, obj.get("level").getAsInt(), obj.get("stat").getAsString());
    }
    
    @Override
    public Identifier getId() {
        return ID;
    }
    
    public void trigger(ServerPlayerEntity player) {
        this.trigger(player, (levelCriteria) -> levelCriteria.matches(player));
    }
    
    public static class LevelCriteria extends AbstractCriterionConditions {
        private final int level;
        private final Identifier id;
        
        public LevelCriteria(EntityPredicate.Extended playerPredicate, int lvl, String id) {
            super(ID, playerPredicate);
            this.level = lvl;
            this.id = new Identifier(id);
        }
        
        public boolean matches(ServerPlayerEntity player) {
            if (id.equals(ANY_ID)) {
                return RPGStats.getHighestLevel(player) >= level;
            } else {
                return CustomComponents.STATS.get(player).getOrCreateID(id).getLevel() >= level;
            }
        }
        
        @Override
        public JsonObject toJson(AdvancementEntityPredicateSerializer ser) {
            JsonObject jsonObject = super.toJson(ser);
            jsonObject.addProperty("level", level);
            jsonObject.addProperty("stat", id.toString());
            return jsonObject;
        }
    }
}

File: src/main/kotlin/io/github/silverandro/rpgstats/advancemnents/LevelUpCriterion.kt
/*
 *   This Source Code Form is subject to the terms of the Mozilla Public
 *   License, v. 2.0. If a copy of the MPL was not distributed with this
 *   file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package io.github.silverandro.rpgstats.advancemnents

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import io.github.silverandro.rpgstats.Constants.ANY_ID
import io.github.silverandro.rpgstats.LevelUtils.getHighestLevel
import io.github.silverandro.rpgstats.advancemnents.LevelUpCriterion.LevelCriteria
import io.github.silverandro.rpgstats.stats.Components
import net.minecraft.advancement.criterion.AbstractCriterion
import net.minecraft.predicate.entity.LootContextPredicate
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.Identifier
import java.util.*

class LevelUpCriterion : AbstractCriterion<LevelCriteria>() {
    fun trigger(player: ServerPlayerEntity) {
        this.trigger(player) { levelCriteria: LevelCriteria -> levelCriteria.matches(player) }
    }

    @JvmRecord
    data class LevelCriteria(val statID: Identifier, val level: Int) : Conditions {
        fun matches(player: ServerPlayerEntity): Boolean {
            return if (statID == ANY_ID) getHighestLevel(player) >= level else Components.STATS.get(player)
                .getOrCreateID(statID).level >= level
        }

        override fun player(): Optional<LootContextPredicate> {
            return Optional.empty()
        }
    }
    override fun getConditionsCodec(): Codec<LevelCriteria> {
        return CODEC
    }

    companion object {
        val CODEC: Codec<LevelCriteria> = RecordCodecBuilder.create { instance ->
            instance.group(
                Identifier.CODEC.fieldOf("statID").forGetter(LevelCriteria::statID),
                Codec.INT.fieldOf("level").forGetter(LevelCriteria::level)
            ).apply(
                instance,
                ::LevelCriteria
            )
        }
    }
}